### PR TITLE
EVG-16585: Update dev tooling

### PR DESCRIPTION
**Pull request by minnakt**
[EVG-16585](https://jira.mongodb.org/browse/EVG-16585)

### Description 
This PR updates `eslint`, `prettier`, and `@graphql-codegen` to latest versions.

Important updates:
- TSLint is deprecated in favor of ESLint. I removed all dependencies and code related to `tslint`.
- We no longer need to specify `prettier/@typescript-eslint` or `prettier/react` because they have been folded into `prettier`. (Specifying `prettier` is enough.)
- All unit test files have been updated so that they no longer deconstruct functions from render e.g. `const { getByDataCy } = render(...)`. Instead they use methods from screen e.g. `screen.getByDataCy`.
- Removed the hacky assertion `await act(() => new Promise((resolve) => setTimeout(resolve, 0)))` that we were using in a couple tests.

Other things:
- Some of the spacing/indentation rules changed so some of the files are literally exactly the same but with different use of spacing.
- Lots of new rules, some of them including:
    - `prefer-regex-literals`: Prevents use of `RegExp` if regex is not dynamically changing. I modified some regex for this rule.
    -  `react/destructuring-assignment`: Forces use of destructuring instead of dot notation. I disabled this entirely because I think it's useful to have dot notation.
    - `react/jsx-no-useless-fragment`: Prevents usage of fragments when they're not needed. However this rule isn't as intelligent as I expected and sometimes we need these so-called "useless" fragments, so I disabled it in the code where applicable.
    - `react/no-unstable-nested-components`: Prevents defining components in `render`. This affects a bunch of different things like `ConditionalWrapper`, so I split it off to another ticket ([EVG-17265](https://jira.mongodb.org/browse/EVG-17265)).
    - `react/jsx-no-constructed-context-values`: Prevents non-stable values from being used as props to `Context.Provider`. I was able to fix this for the `DispatchToast` provider, but ignored it for the `ProjectSettings` provider because it didn't seem like a straightforward fix.

### Important
I think `check-codegen` broke on main after we updated `graphql`. It always shows success even though it technically didn't succeed. ([example](https://spruce.mongodb.com/task/spruce_ubuntu1604_check_codegen_487c1c2a61aa45d1039a4a6a613cb5c85c16b749_22_07_11_17_18_21/logs?execution=0))

Screen Shot 2022-07-12 at 10 04 20 AM

However the update for `@graphql-codegen` in this PR should resolve this issue! 

### Screenshots
- No visible changes

### Testing 
- Modified Jest tests

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -201,7 +199,7 @@
     "lint-staged": "10.5.3",
```
[the remark this answers is not in the document]
Comment (sophstad) [reply package.json:199]: Oh, `husky` is another dependency that may fall under the dev tooling umbrella!

File: .eslintrc.js [hunk 2]
```diff
@@ -37,7 +36,7 @@ module.exports = {
     ecmaFeatures: {
       jsx: true,
     },
-    ecmaVersion: 2018,
+    ecmaVersion: "latest",
```
Comment (khelif96) [inline .eslintrc.js:39]: is this the equivalent of esnext?

**Comment by minnakt** [comment]
I updated `husky` but it required some other configuration stuff, I would be grateful if someone could check that the pre-commit hook still works for them 🙂

If it doesn't work I guess try running the command from [here](https://github.com/typicode/husky-4-to-8)

File: src/components/ErrorBoundary/index.tsx [hunk 3]
```diff
@@ -4,12 +4,8 @@ import BugsnagPluginReact from "@bugsnag/plugin-react";
...
 import ErrorFallback from "./ErrorFallback";
 
-const {
-  getBugsnagApiKey,
-  isProductionBuild,
-  getAppVersion,
-  getReleaseStage,
-} = environmentalVariables;
+const { getBugsnagApiKey, isProductionBuild, getAppVersion, getReleaseStage } =
+  environmentalVariables;
```
[the remark this answers is not in the document]
Comment (minnakt) [reply src/components/ErrorBoundary/index.tsx:8]: I'm guessing `prettier` just has new rules for spacing/indentation

Screen Shot 2022-07-12 at 12 40 59 PM

File: .eslintrc.js (hunk 2 above)
Comment (minnakt) [reply .eslintrc.js:39]: I think so 🤔 ([docs](https://eslint.org/docs/latest/user-guide/configuring/language-options))

Screen Shot 2022-07-12 at 12 42 34 PM

**Comment by minnakt** [comment]
Hmm I might have to push `typescript` down to `4.6` since `storybook` is not happy

I think it's something to do with `ts-node `which we don't have directly installed but it's in `yarn.lock`

**Comment by khelif96** [comment]
> Hmm I might have to push `typescript` down to `4.6` since `storybook` is not happy. I think it might have something to do with `ts-node` which we don't have directly installed but it's listed in `yarn.lock`.

Does installing/pinning a newer version of ts-node work? if not downgrading to 4.6 SGTM

**Review by khelif96 — APPROVED** [review]
mod the typescript changes

**Review by sophstad — APPROVED** [review]
It may be preferable to use [selective dependency resolution](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/) to override the version of `ts-node` since it is not really a spruce dependency. Something like this:
```json
"resolutions": {
    "@graphql-codegen/cli/**/ts-node": "10.8.2"
  }
```

**Pull request merged**